File: C#/Update an Item/UpdateItem.cs
using System;
using System.IO;
using System.Net;
using System.Collections.Generic;

// Newtonsoft.Json is available from http://www.newtonsoft.com/json
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

public class Searching
{
    static private string baseURL = "{base url}/rest/latest/";

    // Username and password should be stored according
    // to your organization's security policies
    static private string username = "API_User";
    static private string password = "********";

    static private string stringToFind = "Unique string";

    static public void Main()
    {
        long itemId = getId(stringToFind);
        updateItem(itemId);
    }

    static public JObject get(string requestURL) 
    {
        HttpWebRequest request = (HttpWebRequest)WebRequest.Create(requestURL);
        request.Credentials = new NetworkCredential(username, password);
        request.Method = WebRequestMethods.Http.Get;
        request.PreAuthenticate = true;

        HttpWebResponse response = (HttpWebResponse)request.GetResponse();

        Stream stream = response.GetResponseStream();
        StreamReader streamReader = new StreamReader(stream);

        return JObject.Parse(streamReader.ReadToEnd());
    }

    static public int put(string requestURL, JObject payload)
    {
        HttpWebRequest request = (HttpWebRequest) WebRequest.Create(requestURL);
        request.Credentials = new NetworkCredential(username, password);
        request.Method = WebRequestMethods.Http.Put;
        request.ContentType = "application/json; charset=UTF-8";
        request.PreAuthenticate = true;

        StreamWriter streamWriter = new StreamWriter(request.GetRequestStream());
        streamWriter.Write(payload.ToString());
        streamWriter.Flush();

        HttpWebResponse response = (HttpWebResponse) request.GetResponse();

        return (int)response.StatusCode;
    }

    static public long getId(string toFind)
    {
        string url = baseURL + "abstractitems?contains=" + WebUtility.UrlEncode(stringToFind);
        JObject response = get(url);
        JObject meta = (JObject) response["meta"];
        JObject pageInfo = (JObject) meta["pageInfo"];
        long totalResults = (long) pageInfo["totalResults"];

        if(totalResults == 1)
        {
            JArray data = (JArray) response["data"];
            JObject item = (JObject) data[0];

            return (long) item["id"];
        }
        else
        {
            Console.WriteLine("stringToFind wasn't unique");
            Environment.Exit(1);
        }
        return -1;
    }

    static public void updateItem(long itemId)
    {
        setLockState(true, itemId);
        JObject item = getItem(itemId);
        item["fields"]["description"] += testResults();
        putItem(itemId, item);
        setLockState(false, itemId);
    }

    static public void setLockState(bool locked, long itemId)
    {
        JObject payload = new JObject();    
        payload.Add("locked", locked);

        string url = baseURL + "items/" + itemId + "/lock";
        put(url, payload);
    }

    static public JObject getItem(long itemId)
    {
        string url = baseURL + "items/" + itemId;
        JObject jsonResponse = get(url);
        return (JObject) jsonResponse["data"];
    }

    static public void putItem(long itemId, JObject item)
    {
        string url = baseURL + "items/" + itemId;

        int responseCode = put(url, item);
        if(responseCode < 400)
        {
            Console.WriteLine("Success");
        }
    }

    static public string testResults()
    {
        Random random = new Random();
        string htmlTemplate = "<p><strong>Imported test results:</strong></p>Status:&nbsp;";

        bool testPassed = random.Next(2) == 1;

        if(testPassed)
        {
            return htmlTemplate + "pass";
        }
        return htmlTemplate + "fail";
    }
}

File: PHP/Update an Item/updateItem.php
<?php
$base_url = "{base url}/rest/latest/";

# Username and password should be stored according
# to your organization's security policies
$username = "API_User";
$password = "********";

$string_to_find = "Unique string";

$item_id = getId($string_to_find);
update_item($item_id);

function get($url) {
    global $username, $password;

    $curl_request = curl_init();
    curl_setopt($curl_request, CURLOPT_URL, $url);
    curl_setopt($curl_request, CURLOPT_USERPWD, "$username:$password");
    curl_setopt($curl_request, CURLOPT_RETURNTRANSFER, True);

    $result = curl_exec($curl_request);
    curl_close($curl_request);

    return json_decode($result, true);
}

function put($url, $payload) {
    global $username, $password;

    $curl_request = curl_init();
    curl_setopt($curl_request, CURLOPT_URL, $url);
    curl_setopt($curl_request, CURLOPT_USERPWD, "$username:$password");
    curl_setopt($curl_request, CURLOPT_RETURNTRANSFER, True);
    curl_setopt($curl_request, CURLOPT_CUSTOMREQUEST, 'PUT');
    curl_setopt($curl_request, CURLOPT_POSTFIELDS, json_encode($payload));
    curl_setopt($curl_request, CURLOPT_HTTPHEADER, array('Content-Type: application/json'));
    curl_setopt($curl_request, CURLOPT_RETURNTRANSFER, True);

    curl_exec($curl_request);
    $response_code = curl_getinfo($curl_request, CURLINFO_HTTP_CODE);
    curl_close($curl_request);
    return $response_code;
}

function getId($to_find) {
    global $base_url;

    $url = $base_url . "abstractitems?contains=" . urlencode($to_find);
    $json_response = get($url);

    $total_results = $json_response['meta']['pageInfo']['totalResults'];

    if($total_results == 1) {
        $data = $json_response['data'];
        $item = $data[0];
        return $item['id'];
    }
    else {
        echo "string_to_find wasn't unique";
        exit(1);
    }
}

function update_item($item_id) {
    set_lock_state(True, $item_id);
    $item = get_item($item_id);
    $item['fields']['description'] .= test_results();
    put_item($item_id, $item);
    set_lock_state(False, $item_id);
}

function set_lock_state($new_lock_state, $item_id) {
    global $base_url;

    $payload = array('locked' => $new_lock_state);
    $url = $base_url . "items/" . $item_id . "/lock";   
    put($url, $payload);
}

function get_item($item_id) {
    global $base_url;

    $url = $base_url . "items/" . $item_id;
    $json_response = get($url);

    return $json_response['data'];
}

function put_item($item_id, $item) {
    global $base_url;

    $url = $base_url . "items/" . $item_id;
    $status_code = put($url, $item);
    if($status_code < 400) {
        echo "Success";
    }
}

function test_results() {
    $template = "<p><strong>Imported test results:</strong></p>Status:&nbsp;";
    $test_passed = rand(0, 1);
    if($test_passed == 1) {
        return $template . "pass";
    }
    return $template . "fail";
}


?>
